Src/AutoFixture.xUnit.net.UnitTest/FrozenAttributeTest.cs

```diff
@@ -82,18 +69,129 @@ public void GetCustomizationWithSpecificRegisteredTypeReturnsCorrectResult()
+            Assert.NotNull(exactTypeMatcher);
+            Assert.Equal(parameter.ParameterType, exactTypeMatcher.TargetType);
+            // Teardown
+        }
+
+        [Fact]
+        public void GetCustomizationShouldMatchBySeedRequestForParameterType()
+        {
+            // Fixture setup
+            var parameter = AParameter<object>();
+            var sut = new FrozenAttribute();
+            // Exercise system
+            var customization = sut.GetCustomization(parameter);
+            // Verify outcome
+            var freezer = Assert.IsAssignableFrom<FreezeOnMatchCustomization>(customization);
+            var matcher = Assert.IsType<OrRequestSpecification>(freezer.Matcher);
+            var seedRequestMatcher = matcher.Specifications.OfType<SeedRequestSpecification>().SingleOrDefault();
+            Assert.NotNull(seedRequestMatcher);
+            Assert.Equal(parameter.ParameterType, seedRequestMatcher.TargetType);
+            // Teardown
+        }
+
+        [Fact]
+        public void GetCustomizationWithMatchingByDirectBaseTypeShouldMatchByBaseType()
+        {
+            // Fixture setup
+            var sut = new FrozenAttribute(Matching.DirectBaseType);
+            // Exercise system
+            var customization = sut.GetCustomization(AParameter<object>());
+            // Verify outcome
+            var freezer = Assert.IsAssignableFrom<FreezeOnMatchCustomization>(customization);
+            var matcher = Assert.IsType<OrRequestSpecification>(freezer.Matcher);
+            Assert.NotEmpty(matcher.Specifications.OfType<DirectBaseTypeSpecification>());
+            // Teardown
+        }
+
+        [Fact]
+        public void GetCustomizationWithMatchingByImplementedInterfacesShouldMatchByImplementedInterfaces()
+        {
+            // Fixture setup
+            var sut = new FrozenAttribute(Matching.ImplementedInterfaces);
+            // Exercise system
+            var customization = sut.GetCustomization(AParameter<object>());
+            // Verify outcome
+            var freezer = Assert.IsAssignableFrom<FreezeOnMatchCustomization>(customization);
+            var matcher = Assert.IsType<OrRequestSpecification>(freezer.Matcher);
+            Assert.NotEmpty(matcher.Specifications.OfType<ImplementedInterfaceSpecification>());
+            // Teardown
+        }
+
+        [Fact]
+        public void GetCustomizationWithMatchingByParameterNameShouldMatchByParameter()
+        {
+            // Fixture setup
+            var sut = new FrozenAttribute(Matching.ParameterName, "parameter");
+            // Exercise system
+            var customization = sut.GetCustomization(AParameter<object>());
+            // Verify outcome
+            var freezer = Assert.IsAssignableFrom<FreezeOnMatchCustomization>(customization);
+            var matcher = Assert.IsType<OrRequestSpecification>(freezer.Matcher);
+            Assert.NotEmpty(matcher.Specifications.OfType<ParameterSpecification>());

```

Comment:
Yes, the test doesn't exercise `IEquatable<ParameterInfo>` so it would definitely be the case.

Let's see where we are at. Since `ParameterSpecification` no longer contains any information about the target of the comparison – instead delegating the operation to an external function – we can't assert on the state of `parameterSpecification`.

At this point we have at least two options:

1. Exercise the `IEquatable<ParameterInfo>` used by `parameterSpecification` and assert that it returns `true` when passed a matching parameter. For example:
    
    ```csharp
    Assert.True(parameterSpecification.IsSatisfiedBy(parameter));
    ```

2. Delete the test and rely solely on the [end-to-end test](https://github.com/AutoFixture/AutoFixture/blob/freeze-attribute/Src/AutoFixture.xUnit.net.UnitTest/Scenario.cs#L182-L188) in `Scenario`.

Going with option **1** would increase the complexity of the test for `PropertySpecification` and `FieldSpecification` because we would have to create both a `ParameterInfo` _and_ a matching `PropertyInfo` and `FieldInfo`. Also, it would no longer be a _unit_ test since we would be acting on multiple objects, namely the `FrozenAttribute` and the implementation of `IEquatable<T>` used by the specification.

Based on that, I suggest we drop the test and simply rely on the end-to-end test in `Scenario`.